### ISSUE TITLE
error with `--no-cache` in composer 1.9

### ISSUE THREAD

**Issue by sudomabider** [hirak/prestissimo#199]
This may or may not be an actual issue. But composer 1.9 now comes with a `--no-cache` global option which I tried to use for installing prestissimo in docker so that I don't have to call `composer clear-cache` afterwards. I understand the 2 are different but I thought perhaps prestissimo won't do anything yet when it's first installed (excuse my lack of understanding of how composer plugins work).

`composer global require --no-cache hirak/prestissimo -v` gives

```
/app # composer global require --no-cache hirak/prestissimo -v                                                                                                                                                                        Changed current directory to /tmp
Using version ^0.3.9 for hirak/prestissimo
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
Dependency resolution completed in 0.000 seconds
Analyzed 75 packages to resolve dependencies
Analyzed 46 rules to resolve dependencies
Package operations: 1 install, 0 updates, 0 removals                                                                                                                                                                                  Installs: hirak/prestissimo:0.3.9
  - Installing hirak/prestissimo (0.3.9): Downloading (100%)
 Extracting archivePlugin installation failed, rolling back
  - Removing hirak/prestissimo (0.3.9)

Installation failed, deleting ./composer.json.                                                                                                                                                                                        

  [ErrorException]
  mkdir(): Not a directory  

Exception trace:                                                                                                                                                                                                                       () at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:103
Composer\Util\ErrorHandler::handle() at n/a:n/a                                                                                                                                                                                       mkdir() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:103
Hirak\Prestissimo\CopyRequest->createDir() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:89
Hirak\Prestissimo\CopyRequest->setDestination() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:39
Hirak\Prestissimo\CopyRequest->__construct() at /tmp/vendor/hirak/prestissimo/src/ParallelizedComposerRepository.php:26
Hirak\Prestissimo\ParallelizedComposerRepository->preloadProviderListings() at /tmp/vendor/hirak/prestissimo/src/ParallelizedComposerRepository.php:39
Hirak\Prestissimo\ParallelizedComposerRepository->prefetch() at /tmp/vendor/hirak/prestissimo/src/Plugin.php:150
Hirak\Prestissimo\Plugin->prefetchComposerRepositories() at /tmp/vendor/hirak/prestissimo/src/Plugin.php:82
Hirak\Prestissimo\Plugin->activate() at phar:///usr/bin/composer/src/Composer/Plugin/PluginManager.php:237
Composer\Plugin\PluginManager->addPlugin() at phar:///usr/bin/composer/src/Composer/Plugin/PluginManager.php:206
Composer\Plugin\PluginManager->registerPackage() at phar:///usr/bin/composer/src/Composer/Installer/PluginInstaller.php:62
Composer\Installer\PluginInstaller->install() at phar:///usr/bin/composer/src/Composer/Installer/InstallationManager.php:173
Composer\Installer\InstallationManager->install() at phar:///usr/bin/composer/src/Composer/Installer/InstallationManager.php:160
Composer\Installer\InstallationManager->execute() at phar:///usr/bin/composer/src/Composer/Installer.php:597
Composer\Installer->doInstall() at phar:///usr/bin/composer/src/Composer/Installer.php:229
Composer\Installer->run() at phar:///usr/bin/composer/src/Composer/Command/RequireCommand.php:218
Composer\Command\RequireCommand->doUpdate() at phar:///usr/bin/composer/src/Composer/Command/RequireCommand.php:175
Composer\Command\RequireCommand->execute() at phar:///usr/bin/composer/vendor/symfony/console/Command/Command.php:245
Symfony\Component\Console\Command\Command->run() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:835
Symfony\Component\Console\Application->doRunCommand() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:185
Symfony\Component\Console\Application->doRun() at phar:///usr/bin/composer/src/Composer/Console/Application.php:267
Composer\Console\Application->doRun() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:117
Symfony\Component\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Console/Application.php:106
Composer\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Command/GlobalCommand.php:101
Composer\Command\GlobalCommand->run() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:835
Symfony\Component\Console\Application->doRunCommand() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:185
Symfony\Component\Console\Application->doRun() at phar:///usr/bin/composer/src/Composer/Console/Application.php:267
Composer\Console\Application->doRun() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:117
Symfony\Component\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Console/Application.php:106
Composer\Console\Application->run() at phar:///usr/bin/composer/bin/composer:61
require() at /usr/bin/composer:24
```

Works fine without the `--no-cache` option. Then trying to install any package with the `--no-cache` option produces the same error which kind of makes sense as prestissimo relies on a cache directory. I guess a more friendly behavior is to fall back to normal composer if cache is disabled. But it's not a biggie.

The question is whether the cache directory is a hard requirement for installing prestissimo for the first time?

**Comment by rihardssceredins** [comment]
I'm having the same issue after upgrading to composer 1.9.0.
If I add the new `--no-cache` option to `composer` command in my Dockerfile like this:

```
RUN composer global require hirak/prestissimo
RUN composer install -vvv \
    --no-progress \
    --no-interaction \
    --ignore-platform-reqs \
    --prefer-dist \
    --no-scripts \
    --no-dev \
    --no-autoloader \
    --no-cache
```

I get this error:

```
Disabling cache usage
Reading ./composer.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/app): git branch --no-color --no-abbrev -v
Executing command (/app): git describe --exact-match --tags
Executing command (/app): git log --pretty="%H" -n1 HEAD
Executing command (/app): hg branch
Executing command (/app): fossil branch list
Executing command (/app): fossil tag list
Executing command (/app): svn info --xml
Reading /tmp/composer.json
Loading config file /tmp/composer.json
Reading /tmp/vendor/composer/installed.json
Loading plugin Hirak\Prestissimo\Plugin
Running 1.9.0 (2019-08-02 20:55:32) with PHP 7.3.8 on Linux / 4.9.184-linuxkit
Reading ./composer.lock
Loading composer repositories with package information
Installing dependencies from lock file
Reading ./composer.lock
Resolving dependencies through SAT
Looking at all rules.

Dependency resolution completed in 0.001 seconds

                            
  [ErrorException]          
  mkdir(): Not a directory  
                            

Exception trace:
 () at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:103
 Composer\Util\ErrorHandler::handle() at n/a:n/a
 mkdir() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:103
 Hirak\Prestissimo\CopyRequest->createDir() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:89
 Hirak\Prestissimo\CopyRequest->setDestination() at /tmp/vendor/hirak/prestissimo/src/CopyRequest.php:39
 Hirak\Prestissimo\CopyRequest->__construct() at /tmp/vendor/hirak/prestissimo/src/Prefetcher.php:83
 Hirak\Prestissimo\Prefetcher->fetchAllFromOperations() at /tmp/vendor/hirak/prestissimo/src/Plugin.php:168
 Hirak\Prestissimo\Plugin->onPostDependenciesSolving() at n/a:n/a
 call_user_func() at phar:///usr/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:176
 Composer\EventDispatcher\EventDispatcher->doDispatch() at phar:///usr/bin/composer/src/Composer/EventDispatcher/EventDispatcher.php:135
 Composer\EventDispatcher\EventDispatcher->dispatchInstallerEvent() at phar:///usr/bin/composer/src/Composer/Installer.php:491
 Composer\Installer->doInstall() at phar:///usr/bin/composer/src/Composer/Installer.php:229
 Composer\Installer->run() at phar:///usr/bin/composer/src/Composer/Command/InstallCommand.php:122
 Composer\Command\InstallCommand->execute() at phar:///usr/bin/composer/vendor/symfony/console/Command/Command.php:245
 Symfony\Component\Console\Command\Command->run() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:835
 Symfony\Component\Console\Application->doRunCommand() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:185
 Symfony\Component\Console\Application->doRun() at phar:///usr/bin/composer/src/Composer/Console/Application.php:267
 Composer\Console\Application->doRun() at phar:///usr/bin/composer/vendor/symfony/console/Application.php:117
 Symfony\Component\Console\Application->run() at phar:///usr/bin/composer/src/Composer/Console/Application.php:106
 Composer\Console\Application->run() at phar:///usr/bin/composer/bin/composer:61
 require() at /usr/bin/composer:24
```

**Comment by FantomX1** [comment]
It's somehow a standard behavior due to composer documentation "Disables the use of the cache directory. Same as setting the COMPOSER_CACHE_DIR env var to /dev/null", prestissimo relies on composer cache dir even when it's run with no cache, and I debugged it via composer github source code, and it failed at prestissimo class in root CopyRequest.php method createDir, line 103
```
 private function createDir($fileName)
    {
        $targetdir = dirname($fileName);
        if (!file_exists($targetdir)) {
            if (!mkdir($targetdir, 0775, true)) {
                throw new FetchException(
                    'The file could not be written to ' . $fileName
                );
            }
        }
    }
```
target dir is in case of supplying --no-cache param set to '/dev/null/something/long/path'

**Comment by polarathene** [comment]
> sometimes even dependencies have bugs, and if you then force using always the same, then you have a sort of problem if those with the same version number

@FantomX1 so installing other packages even without Prestissimo installed can also fail with `--no-cache` for similar reasons? I guess it's better to just clear cache afterwards then? Or was this about bugs due to prior cached packages?

Any reason to want to prefer `--no-cache` at least for Dockerfiles where it really shouldn't make much of a difference/hassle?

**Comment by FantomX1-github** [comment]
@polarathene , I wanted to know, if the bug was not only in cache, as on cacheless docker Continuous Integration reverting dependencies to older composer.lock did help, but it did not help at my composer caching local docker. It my sound like, if revert composer.lock did not help, erasing cache would not help either as it would have save version number,  but I believe lot of developers dont increase version number with each git commit, so the very same versions might behave differently at different times. Actually erasing cache did help me. I don't think also other dependencies must have this issue, firstly I just think majority of dependencies don't utilize cache, this is not necesarilly prestissimo's bug, but a bug in contract which composer introduced in new version, perhaps didn;t realize some dependency might utilize internal composer cache core.

**Comment by akondas** [comment]
The question is this: can we decide on one solution, and I would be happy to fix it. 
@hirak what do you think about it?

**Comment by ThaDaVos** [comment]
Any progress on a solution? Also getting the `mkdir()` error when running `composer install --no-cache` in a project

**Comment by ThaDaVos** [comment]
I have on solution -> if `--no-cache` is supplied, disable prestissimo
Or if possible, do not download to cache but to the vendor folder instead maybe, or wherever the packages go when using `--no-cache`.

**Comment by 4n70w4** [comment]
The same problem!

**Comment by Verhaeg** [comment]
One option here is when checking `--no-cache` instead of not using cache (as this is the core solution of prestissimo). Create a temporary folder instead to be used as cache.
For what it seems, Prestissimo downloads in parallel all packages and then lets composer install them from cache. 

Not sure if feasible as it would need to overwrite the --no-cache flag to use a tmp folder and use cache again, but would be a nice solution instead of disabling this package.

**Comment by edutrul** [comment]
Same having this issue 
![image](https://user-images.githubusercontent.com/1582129/80923941-17875f00-8d4c-11ea-8210-eb10e0515263.png)

Subscribing.

**Comment by edutrul** [comment]
Alright, so the fix here https://github.com/hirak/prestissimo/pull/214/files works as expected. I applied manually and worked. Though it may be better that PR gets other review then merged. Since this is a blocker

**Comment by FantomX1** [comment]
A workaround is also using a command 'composer clear-cache' however it might be not very desired, as it affects the global state, not just the currently running command.

**Comment by GrzegorzDrozd** [comment]
Hi,

Any progress on this issue? 

Have a nice day.
